### ISSUE TITLE
cross-build for arm64

### ISSUE THREAD

**Issue by vielmetti** [daniel-kun/kube-alive#2]
I'm here from https://github.com/docker/cli/issues/327#issuecomment-362732718 and https://dockercommunity.slack.com/archives/C2293P89Y/p1517612020000352

I'm interested in what it would take to also cross-build for arm64. You mention in the build.sh script that there's a dependency on images that have "docker base images to the new ones that include elm for amd64 and arm32v7".  Are there any other components that you'd expect to need? (something about qemu)

I have a native arm64 system that I can push a build from and I'm interested in this as one way to addresss a more general build pattern.

The end goal is to test scaling on the Packet Type 2A (96-core arm64 Cavium ThunderX) system - with 96 cores it should have some interesting properties.

**Comment by daniel-kun** [comment]
Cross-building for arm64 (is that aarch64 then?) in general would require two things:
1) a docker image with linux for arm64 *and* /usr/bin/qemu-aarch64-static *for x86* - I'll call these "qemu-arm64-images" from now on
2) [binfmt_misc](https://en.wikipedia.org/wiki/Binfmt_misc) set up to handle arm64 binaries with qemu-aarch64-static. You can use [this docker image](https://hub.docker.com/r/multiarch/qemu-user-static/) for that.
This blog post should help you:
https://blog.hypriot.com/post/setup-simple-ci-pipeline-for-arm-images/
And here's something similar:
http://www.ecliptik.com/Cross-Building-and-Running-Multi-Arch-Docker-Images/

For cross-building kube-alive (which would be _awesome_), some more things are needed:
1) An "qemu-arm64-image" with Go and GOPATH set to /go (you can probably base this off of [arm64v8/golang](https://hub.docker.com/r/arm64v8/golang/) (I'm really confused by aarch64 vs. arm64)
2)  An "qemu-arm64-image" with nginx and Elm installed. This one is tricky, since you'll need to build the Elm binaries yourself. Luckily, I described how to do this on Raspberry Pi [here](https://dev.to/danielkun/using-elm-on-a-raspberry-pi-42pi). Should be very similar on arm64. And you can use the [official nginx image](https://hub.docker.com/r/arm64v8/nginx/) as a base I guess. Just copy over qemu-aarch64-static and elm-make  built for arm64.
3) Tweak build.sh to add the new arch and use the appropriate base images.

This should get you going. To actually deploy, I'd first replace the original images with your own image in the Kubernetes manifests in the deploy/ folder. You can even use a local registry for this then, which should be handy.

Good look! And let me know when you run into problems, I'll try my best to help you out.

**Comment by deitch** [comment]
Yeah, arm64 = aarch64. You should see some Makefiles and build scripts I have modified to have:

```Makefile
ifeq ($(ARCH),aarch64)
        ARCH=arm64
endif
ifeq ($(ARCH),x86_64)
        ARCH=amd64
endif
```

Oh well.

> Cross-building for arm64 ...

I have been going through that process slowly for Project Calico. Question I have is why you need qemu and binfmt if you are doing purely go binaries? I can run `GOARCH=aarch64 go build .` on an amd64 machine without needing qemu. What do I gain by running it in qemu?

**Comment by daniel-kun** [comment]
@deitch Indeed I could use multi-stage Dockerfiles and cross-build Go binaries in the first stage and copy them to the actual target-platform in the second stage. That's a good idea, I think I'll change the build process to that in the future. (I just recently learned about multi-stage docker builds)

@vielmetti Did you make progress regarding cross-compilation for arm64? Since I like the idea of @deitch , I might not cross-build using qemu in the future and hence will not have the need to cross-build for arm64 to make kube-alive run on arm64. (However, I will still need amd64 hardware to build a few base containers that include things like docker and kubectl for my newer experiments that I'm currently implementing).

**Comment by deitch** [comment]
@daniel-kun I have been using exactly that to help bootstrap calico into the arm64 world. If you look at these three PRs, it should give an idea:

* https://github.com/projectcalico/bird/pull/52
* https://github.com/projectcalico/go-build/pull/26
* https://github.com/projectcalico/felix/pull/1703

**Comment by vielmetti** [comment]
@daniel-kun - I have been working with @deitch on this and he knows a bit more than I know.  If you have experiments that you'd like to do that would have use for fast arm64 hardware, the Works on Arm cluster might very well be the place to do that testing. Drop me a note and I can walk you through getting access.

**Comment by daniel-kun** [comment]
@vielmetti Thanks for the offer. I will finish the current experiment that I'm working on (rolling updates) and will get back to you.
In the meantime, did you get kube-alive cross-build for arm64?
I've recently pushed large changes to the build infrastructure, so if you have changes lying around, its likely that they will be in conflict on the next pull.

**Comment by daniel-kun** [comment]
@vielmetti @deitch I just pushed a fourth experiment (rolling updates without interruptions) yesterday. I will take on a UI refresh based on material next, and could at the same time have a look at cross-building to arm64. If you could send me the steps to get access to the arm64 hardware via e-mail or Twitter DM, that would be great!

**Comment by deitch** [comment]
@daniel-kun hop on the Packet community slack, and message us. I can run a thing or two for you, but the real man for getting access is Ed. Hop over to https://slack.packet.net.

Ed is `@ed.packet` and I am `@deitcher`

**Comment by daniel-kun** [comment]
Fixed with f992fdb3896d4d7ad15db60e617f30fd95474d0b - thanks to packet.net for sponsoring the required hardware!
